compiler/rustc_mir/src/transform/inline.rs

```diff
@@ -115,7 +124,10 @@ impl Inliner<'tcx> {
                     // in the other direction.
                     if !self.tcx.dep_graph.is_fully_enabled()
                         && self_hir_id < callee_hir_id
-                        && caller_body.generator_kind.is_none()
+                        // Or we know for sure that the function we're calling does not recurse.
+                        || !self.tcx.mir_inliner_cycle(callee_def_id)
+                        // Or we know for sure that we don't recurse.
+                        || !self_recurses
```

Comment:
shall or be used here? feels strange that it is ok to inline if callee is not recursive and self is recursive

Reply:
we may be recursive at a different call site. So if you have

```rust
fn foo() {
    a();
    b();
}
fn b() {
    foo();
}
fn a() {}
```

then you can inline the `a` call. This is a bit overeager, as we could inline `a()` even if it were `fn a() { c() }` where `fn c() { a() }`, since there's no cycle going back to `foo`, but I consider that to be a marginal enough use case to ignore it (it's also expensive to check for this)